Int Overflow Bug Reoccurs - Please Add Better Testing

This is the second time in less than two months that an issue with integer overflow happens, breaking builds on 32-bit platforms.  Given that this is the second occurrence of such a bug, I think testing needs to be improved to catch this, because if it has already happened twice in 2 months, it will probably happen a 3rd, and a 4th if testing isn't improved to catch this.

Previous bug report https://github.com/dgraph-io/badger/issues/1142, commit fixing the first time the bug happened https://github.com/dgraph-io/badger/commit/fb0cdb858c78004b9db8f08e91bb713780b20445, and the most recent commit to fix the same issue https://github.com/dgraph-io/badger/commit/465f28aba5fee9a4dac132ccbebaaae6a1d711dd

I've uploaded badger to Debian a while ago, which means it gets built on a number of different architectures. Full logs available here (click red/green status column for full build logs):
https://buildd.debian.org/status/package.php?p=badger

As can be seen by cherry-picking the already committed fix, badger builds on 32bit architectures but testsuite has multiple failures and even crashes and even compilation failures. Maybe it's just the test-suite being broken and badger actually works on 32bit, but I'm scared to just disable the testsuite on all 32bit archs. Would be great if the test-suite also worked on 32 bit archs, so actually running it in your CI would IMHO be nice.


Here are three examples of failures:

```
src/github.com/dgraph-io/badger/db2_test.go:313:14: len argument too large in make([]byte)

$ cat -n db2_test.go  | grep '^ *313'
   313					v := make([]byte, 2<<30)
```
```
=== RUN   TestRotate
badger 2020/02/03 12:04:18 INFO: All 0 tables opened in 0s
badger 2020/02/03 12:04:19 INFO: Got compaction priority: {level:0 score:1.73 dropPrefix:[]}
badger 2020/02/03 12:04:19 INFO: All 0 tables opened in 0s
badger 2020/02/03 12:04:19 INFO: Replaying file id: 0 at offset: 0
badger 2020/02/03 12:04:19 INFO: Replay took: 16.72µs
--- FAIL: TestRotate (1.16s)
    rotate_test.go:53: 
        	Error Trace:	rotate_test.go:53
        	Error:      	Received unexpected error:
        	            	Map log file. Path=/tmp/badger-test546655184/000000.vlog. Error=cannot allocate memory
        	            	During db.vlog.open
        	            	github.com/dgraph-io/badger/y.Wrapf
        	            		/home/ah/badger-2.0.1/obj-arm-linux-gnueabi/src/github.com/dgraph-io/badger/y/error.go:82
        	            	github.com/dgraph-io/badger.Open
        	            		/home/ah/badger-2.0.1/obj-arm-linux-gnueabi/src/github.com/dgraph-io/badger/db.go:357
        	            	github.com/dgraph-io/badger/badger/cmd.TestRotate
        	            		/home/ah/badger-2.0.1/obj-arm-linux-gnueabi/src/github.com/dgraph-io/badger/badger/cmd/rotate_test.go:52
        	            	testing.tRunner
        	            		/usr/lib/go-1.13/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/lib/go-1.13/src/runtime/asm_arm.s:868
        	Test:       	TestRotate
FAIL
```


```
=== RUN   TestTableIndex/multiple_keys
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x39b050]

goroutine 9 [running]:
testing.tRunner.func1(0x20b2280)
	/usr/lib/go-1.13/src/testing/testing.go:874 +0x360
panic(0x3f7dc8, 0x744668)
	/usr/lib/go-1.13/src/runtime/panic.go:679 +0x194
github.com/dgraph-io/badger/table.(*Table).KeyID(...)
	/home/ah/badger-2.0.1/obj-arm-linux-gnueabi/src/github.com/dgraph-io/badger/table/table.go:500
github.com/dgraph-io/badger/table.TestTableIndex.func2(0x20b2280)
	/home/ah/badger-2.0.1/obj-arm-linux-gnueabi/src/github.com/dgraph-io/badger/table/builder_test.go:85 +0xaac
testing.tRunner(0x20b2280, 0x467044)
	/usr/lib/go-1.13/src/testing/testing.go:909 +0xa8
created by testing.(*T).Run
	/usr/lib/go-1.13/src/testing/testing.go:960 +0x2ac
FAIL	github.com/dgraph-io/badger/table	0.290s
```
(FWIW, I guess the last one crashes because the tbl is dereferenced before `require.NoError(t, err, "unable to open table")` is used and OpenTable probably failed for some reason leaving `tbl==nil`. )

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,6 +7,7 @@ go:
   - tip
 os:
   - osx
+  - linux
 env:
   jobs:
   - GOARCH=386

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,7 +7,6 @@ go:
   - tip
 os:
   - osx
-  - linux
 env:
   jobs:
   - GOARCH=386
@@ -20,10 +19,13 @@ jobs:
   allow_failures:
     - go: tip
   exclude:
-    # Exclude builds for 386 architecture on go 1.11 and 1.12
+    # Exclude builds for 386 architecture on go 1.11, 1.12 and tip
+    # Since we don't want it to run for 32 bit
     - go: "1.11"
       env: GOARCH=386
     - go: "1.12"
+      env: GOARCH=386
+    - go: tip
       env: GOARCH=386
 
 notifications:

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,13 +1,20 @@
 language: go
 
 go:
-    - "1.11"
-    - "1.12"
-    - "1.13"
-    - tip
+  - "1.11"
+  - "1.12"
+  - "1.13"
+  - tip
 os:
-    - osx    
-    
+  - osx
+env:
+  jobs:
+  - GOARCH=386
+  - GOARCH=amd64
+  global:
+  - secure: CRkV2+/jlO0gXzzS50XGxfMS117FNwiVjxNY/LeWq06RKD+dDCPxTJl3JCNe3l0cYEPAglV2uMMYukDiTqJ7e+HI4nh4N4mv6lwx39N8dAvJe1x5ITS2T4qk4kTjuQb1Q1vw/ZOxoQqmvNKj2uRmBdJ/HHmysbRJ1OzCWML3OXdUwJf0AYlJzTjpMfkOKr7sTtE4rwyyQtd4tKH1fGdurgI9ZuFd9qvYxK2qcJhsQ6CNqMXt+7FkVkN1rIPmofjjBTNryzUr4COFXuWH95aDAif19DeBW4lbNgo1+FpDsrgmqtuhl6NAuptI8q/imow2KXBYJ8JPXsxW8DVFj0IIp0RCd3GjaEnwBEbxAyiIHLfW7AudyTS/dJOvZffPqXnuJ8xj3OPIdNe4xY0hWl8Ju2HhKfLOAHq7VadHZWd3IHLil70EiL4/JLD1rNbMImUZisFaA8pyrcIvYYebjOnk4TscwKFLedClRSX1XsMjWWd0oykQtrdkHM2IxknnBpaLu7mFnfE07f6dkG0nlpyu4SCLey7hr5FdcEmljA0nIxTSYDg6035fQkBEAbe7hlESOekkVNT9IZPwG+lmt3vU4ofi6NqNbJecOuSB+h36IiZ9s4YQtxYNnLgW14zjuFGGyT5smc3IjBT7qngDjKIgyrSVoRkY/8udy9qbUgvBeW8=
+
+
 jobs:
   allow_failures:
     - go: tip
@@ -16,18 +23,12 @@ jobs:
     - go: "1.11"
       env: GOARCH=386
     - go: "1.12"
-      env: GOARCH=386  
+      env: GOARCH=386
 
 notifications:
   email: false
   slack:
     secure: X7uBLWYbuUhf8QFE16CoS5z7WvFR8EN9j6cEectMW6mKZ3vwXGwVXRIPsgUq/606DsQdCCx34MR8MRWYGlu6TBolbSe9y0EP0i46yipPz22YtuT7umcVUbGEyx8MZKgG0v1u/zA0O4aCsOBpGAA3gxz8h3JlEHDt+hv6U8xRsSllVLzLSNb5lwxDtcfEDxVVqP47GMEgjLPM28Pyt5qwjk7o5a4YSVzkfdxBXxd3gWzFUWzJ5E3cTacli50dK4GVfiLcQY2aQYoYO7AAvDnvP+TPfjDkBlUEE4MUz5CDIN51Xb+WW33sX7g+r3Bj7V5IRcF973RiYkpEh+3eoiPnyWyxhDZBYilty3b+Hysp6d4Ov/3I3ll7Bcny5+cYjakjkMH3l9w3gs6Y82GlpSLSJshKWS8vPRsxFe0Pstj6QSJXTd9EBaFr+l1ScXjJv/Sya9j8N9FfTuOTESWuaL1auX4Y7zEEVHlA8SCNOO8K0eTfxGZnC/YcIHsR8rePEAcFxfOYQppkyLF/XvAtnb/LMUuu0g4y2qNdme6Oelvyar1tFEMRtbl4mRCdu/krXBFtkrsfUaVY6WTPdvXAGotsFJ0wuA53zGVhlcd3+xAlSlR3c1QX95HIMeivJKb5L4nTjP+xnrmQNtnVk+tG4LSH2ltuwcZSSczModtcBmRefrk=
-
-env:
-  global:
-  - GOOS=darwin GOARCH=386
-  - GOOS=darwin GOARCH=amd64
-  - secure: CRkV2+/jlO0gXzzS50XGxfMS117FNwiVjxNY/LeWq06RKD+dDCPxTJl3JCNe3l0cYEPAglV2uMMYukDiTqJ7e+HI4nh4N4mv6lwx39N8dAvJe1x5ITS2T4qk4kTjuQb1Q1vw/ZOxoQqmvNKj2uRmBdJ/HHmysbRJ1OzCWML3OXdUwJf0AYlJzTjpMfkOKr7sTtE4rwyyQtd4tKH1fGdurgI9ZuFd9qvYxK2qcJhsQ6CNqMXt+7FkVkN1rIPmofjjBTNryzUr4COFXuWH95aDAif19DeBW4lbNgo1+FpDsrgmqtuhl6NAuptI8q/imow2KXBYJ8JPXsxW8DVFj0IIp0RCd3GjaEnwBEbxAyiIHLfW7AudyTS/dJOvZffPqXnuJ8xj3OPIdNe4xY0hWl8Ju2HhKfLOAHq7VadHZWd3IHLil70EiL4/JLD1rNbMImUZisFaA8pyrcIvYYebjOnk4TscwKFLedClRSX1XsMjWWd0oykQtrdkHM2IxknnBpaLu7mFnfE07f6dkG0nlpyu4SCLey7hr5FdcEmljA0nIxTSYDg6035fQkBEAbe7hlESOekkVNT9IZPwG+lmt3vU4ofi6NqNbJecOuSB+h36IiZ9s4YQtxYNnLgW14zjuFGGyT5smc3IjBT7qngDjKIgyrSVoRkY/8udy9qbUgvBeW8=
 
 before_script:
 - go get github.com/mattn/goveralls

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,10 +7,19 @@ go:
     - tip
 os:
     - osx
+ env:
+   - GOARCH=386
+   - GOARCH=amd64    
     
 jobs:
   allow_failures:
     - go: tip
+  exclude:
+    # Exclude builds for 386 architecture on go 1.11 and 1.12
+    - go: "1.11"
+      env: GOARCH=386
+    - go: "1.12"
+      env: GOARCH=386  
 
 notifications:
   email: false

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,10 +6,7 @@ go:
     - "1.13"
     - tip
 os:
-    - osx
- env:
-   - GOARCH=386
-   - GOARCH=amd64    
+    - osx    
     
 jobs:
   allow_failures:
@@ -28,6 +25,8 @@ notifications:
 
 env:
   global:
+  - GOOS=darwin GOARCH=386
+  - GOOS=darwin GOARCH=amd64
   - secure: CRkV2+/jlO0gXzzS50XGxfMS117FNwiVjxNY/LeWq06RKD+dDCPxTJl3JCNe3l0cYEPAglV2uMMYukDiTqJ7e+HI4nh4N4mv6lwx39N8dAvJe1x5ITS2T4qk4kTjuQb1Q1vw/ZOxoQqmvNKj2uRmBdJ/HHmysbRJ1OzCWML3OXdUwJf0AYlJzTjpMfkOKr7sTtE4rwyyQtd4tKH1fGdurgI9ZuFd9qvYxK2qcJhsQ6CNqMXt+7FkVkN1rIPmofjjBTNryzUr4COFXuWH95aDAif19DeBW4lbNgo1+FpDsrgmqtuhl6NAuptI8q/imow2KXBYJ8JPXsxW8DVFj0IIp0RCd3GjaEnwBEbxAyiIHLfW7AudyTS/dJOvZffPqXnuJ8xj3OPIdNe4xY0hWl8Ju2HhKfLOAHq7VadHZWd3IHLil70EiL4/JLD1rNbMImUZisFaA8pyrcIvYYebjOnk4TscwKFLedClRSX1XsMjWWd0oykQtrdkHM2IxknnBpaLu7mFnfE07f6dkG0nlpyu4SCLey7hr5FdcEmljA0nIxTSYDg6035fQkBEAbe7hlESOekkVNT9IZPwG+lmt3vU4ofi6NqNbJecOuSB+h36IiZ9s4YQtxYNnLgW14zjuFGGyT5smc3IjBT7qngDjKIgyrSVoRkY/8udy9qbUgvBeW8=
 
 before_script:

--- a/db2_test.go
+++ b/db2_test.go
@@ -23,6 +23,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"log"
+	"math"
 	"math/rand"
 	"os"
 	"path"
@@ -310,7 +311,7 @@ func TestPushValueLogLimit(t *testing.T) {
 
 		for i := 0; i < 32; i++ {
 			if i == 4 {
-				v := make([]byte, 2<<30)
+				v := make([]byte, math.MaxInt32)
 				err := db.Update(func(txn *Txn) error {
 					return txn.SetEntry(NewEntry([]byte(key(i)), v))
 				})

--- a/levels_test.go
+++ b/levels_test.go
@@ -393,7 +393,7 @@ func TestDiscardTs(t *testing.T) {
 func TestDiscardFirstVersion(t *testing.T) {
 	opt := DefaultOptions("")
 	opt.NumCompactors = 0
-	opt.NumVersionsToKeep = math.MaxUint32
+	opt.NumVersionsToKeep = math.MaxInt32
 	opt.managedTxns = true
 
 	runBadgerTest(t, &opt, func(t *testing.T, db *DB) {

--- a/table/builder_test.go
+++ b/table/builder_test.go
@@ -80,11 +80,11 @@ func TestTableIndex(t *testing.T) {
 			require.NoError(t, err, "unable to write to file")
 
 			tbl, err := OpenTable(f, opt)
+			require.NoError(t, err, "unable to open table")
 			if opt.DataKey == nil {
 				// key id is zero if thre is no datakey.
 				require.Equal(t, tbl.KeyID(), uint64(0))
 			}
-			require.NoError(t, err, "unable to open table")
 
 			// Ensure index is built correctly
 			require.Equal(t, blockCount, len(tbl.blockIndex))

